### PR TITLE
Memory detection v2

### PR DESCRIPTION
**Pull request by AndreasMadsen**
This is a complete rewrite of the memory issue detection algorithm.

* Instead of looking at the memory usage, it looks directly at how much the event loop was blocked due to garbage collection.
* The previous implementation did look at event look blocking to some extent. However, it was not a direct measurement, it was only valid for `node.js < 8`, and only looked at events related to the old space.
* The new version can detect issues with both old and new space. In recent node versions, the old space GC is actually so good that it is rarely an issue.
* The new version indirectly handles highly parallelized gc, by looking at actual durations which will be less for highly parallelized gc.
* The new version, also detects issues with short but very frequent blocking gc. The scavenge event, only takes about 5ms but if it runs 100 times over a second that is still a big issue.

Along with the new algorithm, is a new internal document that describes the model in details and contains the following table that I spent quite a while compiling.

| Task                 | parallel | Incremental | Cocurrent | heap space | Appropiate time | Description          | Trace event                     |
|---------------------:|:--------:|:-----------:|:---------:|:-----------|:----------------|:---------------------|:--------------------------------|
| Scavenge             | √ (v10+) |             |           | new        | ~ 5 ms          | Cleanup nursery      | V8.GCScavenger                  |
| Marking              | √ (v10+) | √ (v4+)     | √ (v8+)   | old        |                 | Finding live objects | V8.GCIncrementalMarking         |
| Finalize Marking     |          |             |           | old        |                 |                      | V8.GCIncrementalMarkingFinalize |
| Sweeping             |          |             | √ (v4+)   | old        |                 | Reclaming dead space |                                 |
| Compaction + Update  | √ (v6+)  |             |           | old        | ~ 10 ms         | Defragmenting memory | V8.GCFinalizeMC                 |
| Array buffer freeing |          |             | √ (v10+)  | unknown    |                 |                      |                                 |

### CODE REVIEW

**Comment by AndreasMadsen** [comment]
@goto-bus-stop Could you review this?

File: analysis/analyse-memory.js [hunk 1]
```diff
@@ -2,46 +2,88 @@
...
+      name: d.name,
+      timeWindow: Math.round((0.5 / 1000) * (d.args.endTimestamp + d.args.startTimestamp)),
+      duration: d.args.endTimestamp - d.args.startTimestamp
+    }))
 
-  // If MSC caused a big delay
-  const mscDelayStat = summary(mscDelay)
+  // Setup a 2d array structure, to hold data for each time window
+  const timeWindows = gcevents.map((d) => d.timeWindow)
+  const TimeWindowMax = Math.max(...timeWindows)
```
Comment (goto-bus-stop) [inline analysis/analyse-memory.js]:
```suggestion
  const timeWindowMax = Math.max(...timeWindows)
```

File: test/generate-process-stat.js [hunk 2]
```diff
@@ -3,17 +3,13 @@
 const xorshift = require('xorshift')
 const MB = Math.pow(1024, 2)
 
-function generateProcessStat (data, noiseLevel, timeSpaceing) {
+function generateProcessStat (data, noiseLevel, timeStreching) {
```
Comment (goto-bus-stop) [inline test/generate-process-stat.js]:
```suggestion
function generateProcessStat (data, noiseLevel, timeStretching) {
```

**Comment by AndreasMadsen** [comment]
@goto-bus-stop Thanks. I merged the PR with the changes you suggested.

**Pull request merged**